<img src=".\images\Logo.png" width=150 align="left" /> <img src=".\images\Logo2.jpg" width=450 align="right" />


# <center><font color= #1e8449 > <b>CU18_Infraestructuras_eventos</font></center>

<font color='steelblue'><b>Citizenlab Data Science Methodology > II - Data Processing Domain </font>
***
> # <font color='steelblue'> <b>05.- Data Collection</font>

<font color='steelblue'>Data Collection is the process to obtain and generate (if required) necessary data to model the problem.</font>

### 03. Obtener datos de geometrías secciones censales en formato json

* Dados los ficheros shp del INE, obtener un solo archivo json con la información georeferenciada
* Los datos están por sección censal, hay que unir los polígonos en distritos censales







<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Settings" data-toc-modified-id="Settings-1"><font color="green">Settings</font></a></span></li><li><span><a href="#Data-Load" data-toc-modified-id="Data-Load-2"><font color="green">Data Load</font></a></span></li><li><span><a href="#ETL-Processes" data-toc-modified-id="ETL-Processes-3">ETL Processes</a></span><ul class="toc-item"><li><span><a href="#Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files" data-toc-modified-id="Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files-3.1">Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files</a></span></li></ul></li><li><span><a href="#Synthetic-Data-Generation" data-toc-modified-id="Synthetic-Data-Generation-4">Synthetic Data Generation</a></span></li><li><span><a href="#Fake-Data-Generation" data-toc-modified-id="Fake-Data-Generation-5">Fake Data Generation</a></span></li><li><span><a href="#Open-Data" data-toc-modified-id="Open-Data-6">Open Data</a></span></li><li><span><a href="#Data-Save" data-toc-modified-id="Data-Save-7"><font color="green">Data Save</font></a></span></li><li><span><a href="#Main-Conclusions" data-toc-modified-id="Main-Conclusions-8"><font color=" #2874a6 ">Main Conclusions</font></a></span></li><li><span><a href="#Main-Actions" data-toc-modified-id="Main-Actions-9"><font color=" #2874a6 ">Main Actions</font></a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Acciones-done" data-toc-modified-id="Acciones-done-9.0.1"><font color=" #2874a6 ">Acciones done</font></a></span></li><li><span><a href="#Acctions-to-perform" data-toc-modified-id="Acctions-to-perform-9.0.2"><font color=" #2874a6 ">Acctions to perform</font></a></span></li></ul></li></ul></li></ul></div>

## <font color='green'>Settings</font>

### Packages to use

- {tcltk} para seleccionar archivos locales de forma interactiva
- {sf} para trabajar con georeferenciación
- {dplyr} para explorar datos


In [1]:
library(sf)
library(dplyr)


Linking to GEOS 3.10.2, GDAL 3.4.2, PROJ 8.2.1; sf_use_s2() is TRUE


Attaching package: 'dplyr'


The following objects are masked from 'package:stats':

    filter, lag


The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union




### Paths

In [2]:
iPath <- "Data/Input/"
oPath <- "Data/Output/"

## <font color='green'>Data Load</font>

If there are more than one input file, make as many sections as files to import.

Instrucciones
- Los ficheros de entrada del proceso están siempre en Data/Input/.   
- Si hay más de un fichero de entrada, se crean tantos objetos iFile_xx y file_data_xx como ficheros de entrada (xx número correlativo con dos dígitos, rellenar con ceros a la izquierda)

#### Archivo de geometrías

<font color='tomato'><b> OPCION A:</b> Seleccionar fichero en ventana para mayor comodidad</font> 

Data load using the {tcltk} package. Ucomment the line if not using this option

In [3]:
# file_data <- tcltk::tk_choose.files(multi = FALSE)

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>



In [4]:
iFile <- "Espana_Seccionado2022_ETRS89H30/SECC_CE_20220101.shp"
file_data <- paste0(iPath, iFile)

if(file.exists(file_data)){
    cat("Se leerán datos del archivo: ", file_data)
} else{
    warning("Cuidado: el archivo no existe.")
}


Se leer<U+00E1>n datos del archivo:  Data/Input/Espana_Seccionado2022_ETRS89H30/SECC_CE_20220101.shp

#### Data file to dataframe

Usar la función adecuada según el formato de entrada (xlsx, csv, json, ...)

In [5]:
data <- st_read(file_data)


Reading layer `SECC_CE_20220101' from data source 
  `/Users/emilio.lcano/academico/gh_repos/__transferencia/citizenlab/CitizenLab-Research-and-Development/casos_urjc/notebooks/II_data_processing/18_infraestructuras/Data/Input/Espana_Seccionado2022_ETRS89H30/SECC_CE_20220101.shp' 
  using driver `ESRI Shapefile'
Simple feature collection with 36382 features and 16 fields
Geometry type: MULTIPOLYGON
Dimension:     XY
Bounding box:  xmin: -1004502 ymin: 3132130 xmax: 1126932 ymax: 4859240
Projected CRS: ETRS89 / UTM zone 30N


Estructura de  los datos:

In [6]:
glimpse(data)

Rows: 36,382
Columns: 17
$ CUSEC    <chr> "0100101001", "0100101002", "0100201001", "0100201002", "0100~
$ CUMUN    <chr> "01001", "01001", "01002", "01002", "01002", "01002", "01002"~
$ CSEC     <chr> "001", "002", "001", "002", "003", "004", "005", "006", "007"~
$ CDIS     <chr> "01", "01", "01", "01", "01", "01", "01", "01", "01", "01", "~
$ CMUN     <chr> "001", "001", "002", "002", "002", "002", "002", "002", "002"~
$ CPRO     <chr> "01", "01", "01", "01", "01", "01", "01", "01", "01", "01", "~
$ CCA      <chr> "16", "16", "16", "16", "16", "16", "16", "16", "16", "16", "~
$ CUDIS    <chr> "0100101", "0100101", "0100201", "0100201", "0100201", "01002~
$ CLAU2    <chr> "01001", "01001", "01002", "01002", "01002", "01002", "01002"~
$ NPRO     <chr> "Araba/<U+00C1>lava", "Araba/<U+00C1>lava", "Araba/<U+00C1>lava", "Araba/<U+00C1>lava", "~
$ NCA      <chr> "Pa<U+00ED>s Vasco", "Pa<U+00ED>s Vasco", "Pa<U+00ED>s Vasco", "Pa<U+00ED>s Vasco", "Pa<U+00ED>s~
$ CNUT0    <chr> "ES", "ES", "ES

Muestra de datos:

In [7]:
slice_head(data.frame(data), n = 5)

CUSEC      CUMUN CSEC CDIS CMUN CPRO CCA CUDIS   CLAU2 NPRO              
1 0100101001 01001 001  01   001  01   16  0100101 01001 Araba/<U+00C1>lava
2 0100101002 01001 002  01   001  01   16  0100101 01001 Araba/<U+00C1>lava
3 0100201001 01002 001  01   002  01   16  0100201 01002 Araba/<U+00C1>lava
4 0100201002 01002 002  01   002  01   16  0100201 01002 Araba/<U+00C1>lava
5 0100201003 01002 003  01   002  01   16  0100201 01002 Araba/<U+00C1>lava
  NCA               CNUT0 CNUT1 CNUT2 CNUT3 NMUN                   
1 Pa<U+00ED>s Vasco ES    2     1     1     Alegr<U+00ED>a-Dulantzi
2 Pa<U+00ED>s Vasco ES    2     1     1     Alegr<U+00ED>a-Dulantzi
3 Pa<U+00ED>s Vasco ES    2     1     1     Amurrio                
4 Pa<U+00ED>s Vasco ES    2     1     1     Amurrio                
5 Pa<U+00ED>s Vasco ES    2     1     1     Amurrio                
  geometry                      
1 MULTIPOLYGON (((539753 4743...
2 MULTIPOLYGON (((539559.7 47...
3 MULTIPOLYGON (((503618.6 47...
4 MULTIPOLYGON (((505948.3 47...
5 MULTIPOLYGON (((499919.5 47...

## ETL Processes

### Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files

Se han importado en el apartado Data Load anterior:

* Geometrías de secciones censales del INE

Incluir apartados si procede para: Extracción de datos (select, filter), Transformación de datos, (mutate, joins, ...). Si es necesario tratar datos perdidos, indicarlo también en NB 09.2

#### Extracción de datos

* Filtrar solo comunidad de Madrid

In [8]:
tdata <- data |> 
    filter(CPRO == "28")

#### Transformación de datos

* Transformar sistema de referencia CRS a 4326 (WGS 84)
* Agrupar datos por distrito
* Unir geometrías

In [9]:
tdata <- tdata |>
    st_transform(4258) |> 
    group_by(CMUN, CDIS) |> 
    summarise(geometry = st_union(geometry)) 

`summarise()` has grouped output by 'CMUN', 'CDIS'. You can override using the
`.groups` argument.


Muestra de primeros datos:

In [10]:
tdata |> tibble() |> slice_head(n = 5)

CMUN CDIS geometry                      
1 001  01   POLYGON ((-3.64502 41.12129...
2 002  01   POLYGON ((-3.503032 40.526,...
3 003  01   POLYGON ((-3.808664 40.8921...
4 004  01   POLYGON ((-4.00197 40.25642...
5 005  01   POLYGON ((-3.361691 40.4762...

## Synthetic Data Generation

Si no aplica: Estos datos no requieren tareas de este tipo.

## Fake Data Generation

Si no aplica: Estos datos no requieren tareas de este tipo.

## Open Data

Los archivos shp se descargaron de la fuente abierta como ficheros comprimidos .zip que se descomprimieron para tener la estructura de carpetas necesaria

## <font color='green'>Data Save</font>

Este proceso, puede copiarse y repetirse en aquellas partes del notebbok que necesiten guardar datos.
Recuerde cambiar las cadenas añadida del fichero para diferenciarlas

<font color='tomato'> Identificamos los datos a guardar</font>

In [11]:
data_to_save <- tdata


<font color='tomato'>Estructura de nombre de archivos:</font>

* Código del caso de uso, por ejemplo "CU_04"
* Número del proceso que lo genera, por ejemplo "_05".
* Número de la tarea que lo genera, por ejemplo "_01"
* En caso de generarse varios ficheros en la misma tarea, llevarán _01 _02 ... después
* Nombre: identificativo de "properData", por ejemplo "_zonasgeo"
* Extensión del archivo

Ejemplo: "CU_04_05_01_01_zonasgeo.json, primer fichero que se genera en la tarea 01 del proceso 05 (Data Collection) para el caso de uso 04 (vacunas)

Importante mantener los guiones bajos antes de proceso, tarea, archivo y nombre

### Proceso 05

In [12]:
caso <- "CU_18"
proceso <- '_05'
tarea <- "_03"
archivo <- ""
proper <- "_distritos_geo"
extension <- ".json"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [13]:
# file_save <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out <- paste0(oPath, file_save)
# write_csv(datos, path_out)

# cat('File saved as: ')
# path_out

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [14]:
file_save <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out <- paste0(oPath, file_save)
st_write(obj = data_to_save, 
         dsn = path_out, 
         driver = "GeoJSON",
         delete_dsn = TRUE)

cat('File saved as: ')
path_out

Deleting source `Data/Output/CU_18_05_03_distritos_geo.json' using driver `GeoJSON'
Writing layer `CU_18_05_03_distritos_geo' to data source 
  `Data/Output/CU_18_05_03_distritos_geo.json' using driver `GeoJSON'
Writing 247 features with 2 fields and geometry type Unknown (any).
File saved as: 

[1] "Data/Output/CU_18_05_03_distritos_geo.json"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [15]:
path_in <- paste0(iPath, file_save)
file.copy(path_out, path_in, overwrite = TRUE)

[1] TRUE

## <font color=' #2874a6 '>Main Conclusions</font>

<font color='steelblue'>List and describe the general conclusions of the analysis carried out​.</font>

### Prerequisites

This working code needs the following conditions:

* For using the interactive selection of file, the {tcltk} package must be installed. It is not needed in production.
* The {sf} and {dplyr} packages must be installed. 
* The data paths `Data/Input` and `Data/Output` must exist (relative to the notebook path)

### Configuration Management

This notebook has been tested with the following versions of R and packages. It cannot be assured that later versions work in the same way:
* R 4.2.2
* tcltk 4.2.2
* dplyr 1.0.10
* sf 1.0.9


### Data structures

#### Objeto `data`

- Hay 247 filas correspondientes a los distritos de la comunidad de madrid, con la información:
    * CMUN
    * CDIS
    * geometría


In [16]:
glimpse(tdata)

Rows: 247
Columns: 3
Groups: CMUN, CDIS [246]
$ CMUN     <chr> "001", "002", "003", "004", "005", "005", "005", "005", "005"~
$ CDIS     <chr> "01", "01", "01", "01", "01", "02", "03", "04", "05", "01", "~
$ geometry <POLYGON [arc_degree]> POLYGON ((-3.64502 41.12129..., POLYGON ((-3~


#### Observaciones generales sobre los datos

* Los datos están por sección censal, y se agrupan por distrito

### Consideraciones para despliegue en piloto

* Los datos de origen fueron obtenidos de https://www.ine.es/ss/Satellite?L=es_ES&c=Page&cid=1259952026632&p=1259952026632&pagename=ProductosYServicios%2FPYSLayout en fichero zip con formato .shp y descomprimido en la carpeta Data/Input
* Si la información geográfica cambia, se debería actualizar el fichero y volver a ejecutar todos los procesos.


### Consideraciones para despliegue en producción

* Se deben crear los procesos ETL en producción necesarios para que los datos de entrada estén actualizados

## <font color=' #2874a6 '>Main Actions</font>

#### <font color=' #2874a6 '>Acciones done</font>

<font color='steelblue'>Indicate the actions that have been carried out in this process</font>

- Se ha tenido que renombrar la carpeta descargada del INE de Espa§a_Seccionado2022_ETRS89H30 a Espana_Seccionado2022_ETRS89H30
- Se ha transformado al sistema de referencia CRS a 4326 (WGS 84)
- Se han agrupado datos por distrito
- Se han unido geometrías


#### <font color=' #2874a6 '>Acctions to perform</font>

<font color='steelblue'>Indicate the actions that must be carried out in subsequent processes</font>

- Se deben asignar distritos censales a cada infraestructura

## <font color='RED'>CODE TO DEPLOY (PILOT)</font>

A continuación se incluirá el código que deba ser llevado a despliegue para producción, dado que se entiende efectúa operaciones necesarias sobre los datos en la ejecución del prototipo

<font color=' red '><b>Description</b></font>

- No hay nada que desplegar en el piloto, ya que estos datos son estáticos o en todo caso cambian con muy poca frecuencia, altamente improbable durante el proyecto.


<font color=' red '><b>CODE</b></font>

In [17]:
# incluir código